src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -191,61 +198,84 @@ protected Solution SetCurrentSolution(Solution solution)
+                onBeforeUpdate: static (oldSolution, newSolution, data) =>
+                {
+                    data.onBeforeUpdate?.Invoke(oldSolution, newSolution);
+                },
                 onAfterUpdate: static (oldSolution, newSolution, data) =>
                 {
+                    data.onAfterUpdate?.Invoke(oldSolution, newSolution);
+
                     // Queue the event but don't execute its handlers on this thread.
                     // Doing so under the serialization lock guarantees the same ordering of the events
                     // as the order of the changes made to the solution.
                     data.@this.RaiseWorkspaceChangedEventAsync(data.kind, oldSolution, newSolution, data.projectId, data.documentId);
-                },
-                (@this: this, transformation, kind, projectId, documentId));
+                });
         }
 
         /// <summary>
-        /// Applies specified transformation to <see cref="CurrentSolution"/>, updates <see cref="CurrentSolution"/> to the new value and raises a workspace change event of the specified kind.
+        /// Applies specified transformation to <see cref="CurrentSolution"/>, updates <see cref="CurrentSolution"/> to
+        /// the new value and performs a requested callback immediately before and after that update.  The callbacks
+        /// will be invoked atomically while while <see cref="_serializationLock"/> is being held.
         /// </summary>
         /// <param name="transformation">Solution transformation.</param>
+        /// <param name="onBeforeUpdate">Action to perform immediately prior to updating <see cref="CurrentSolution"/>.
+        /// The action will be passed the old <see cref="CurrentSolution"/> that will be replaced and the exact solution
+        /// it will be replaced with. The latter may be different than the solution returned by <paramref
+        /// name="transformation"/> as it will have its <see cref="Solution.WorkspaceVersion"/> updated
+        /// accordingly.</param>
         /// <param name="onAfterUpdate">Action to perform once <see cref="CurrentSolution"/> has been updated.  The
         /// action will be passed the old <see cref="CurrentSolution"/> that was just replaced and the exact solution it
         /// was replaced with. The latter may be different than the solution returned by <paramref
-        /// name="transformation"/> as it will have its <see cref="Solution.WorkspaceVersion"/> updated accordingly.
-        /// Updating the solution and invoking <paramref name="onAfterUpdate"/> will happen atomically while <see
-        /// cref="_serializationLock"/> is being held.</param>
-        internal bool SetCurrentSolution<TData>(
+        /// name="transformation"/> as it will have its <see cref="Solution.WorkspaceVersion"/> updated
+        /// accordingly.</param>
+        private protected bool SetCurrentSolution<TData>(
```

Comment:
@jasonmalinowski this is the meat of the change.  A new halper which promise to atomically update the current-solution.  And only if it it going to do that does it run the onBeforeUpdate and onAfterUpdate helpers.

THe former is often for use for clearing out state from the workspace prior to the set.  The after is often for use for issuing events after the change.